java-manta-client-unshaded/src/main/java/com/joyent/manta/client/LazyMantaClient.java

```diff
@@ -0,0 +1,299 @@
+     */
+    private final MantaConnectionFactoryConfigurator connectionConfig;
+
+    /**
+     * MBean supervisor.
+     */
+    private final MantaMBeanSupervisor lazyBeanSupervisor;
+
+    /**
+     * Internal component categories.
+     */
+    private enum MantaClientComponent {
+        /**
+         * The {@link AuthenticationConfigurator} and {@link HttpHelper} both derive their state from
+         * the <a href="https://github.com/joyent/java-manta/USAGE.md#configuration-sections">
+         * Authentication/Authorization Section</a> of the {@link ConfigContext}.
+         */
+        AUTH,
+        /**
+         * The {@link HttpHelper} both derives state from both {@link #AUTH} and the
+         * the <a href="https://github.com/joyent/java-manta/USAGE.md#configuration-sections">
+         * Network Section</a> of the {@link ConfigContext}.
+         */
+        HTTP,
+        /**
+         * The {@link UriSigner} both derives state from {@link #AUTH} but is current a separate component.
+         * This value might go away if it were to be rolled into the {@link AuthenticationConfigurator}.
+         */
+        SIGN,
+    }
+
+    /**
+     * Set of internal components which need to be dynamically reloaded.
+     * Calling {@link #reload()} will cause this set to be filled with all values of the enum.
+     * Getters will subsequently remove their component from the {@link EnumSet} once the new
+     * value has been cached.
+     *
+     * PERHAPS: this should be wrapped in an {@link AtomicReference}?
+     */
+    private volatile EnumSet<MantaClientComponent> reload;
+
+    /**
+     * The instance of the http helper class used to simplify creating requests.
+     */
+    private HttpHelper lazyHttpHelper;
+
+    /**
+     * Instance used to generate Manta signed URIs.
+     */
+    private UriSigner lazyUriSigner;
+
+    /**
+     * Build a client which may or may not be immediately useable based on the provided configuration.
+     * We do not validate this config at all since it might have changed by the time the getters read it.
+     *
+     * @param config the {@link ConfigContext} to track
+     */
+    public LazyMantaClient(final ConfigContext config) {
+        this(config, null);
+    }
+
+    /**
+     * Build a client that tracks a {@link ConfigContext} and may or may not include customized internal HttpClient
+     * components. If provided the {@link MantaConnectionFactoryConfigurator} will be passed on to a
+     * {@link MantaConnectionFactory} when the {@link HttpHelper} is being rebuilt.
+     *
+     * @param config           the {@link ConfigContext} to track
+     * @param connectionConfig customized connection objects to use when creating the {@link HttpHelper}, or null
+     */
+    public LazyMantaClient(final ConfigContext config,
+                           final MantaConnectionFactoryConfigurator connectionConfig) {
+        super();
+        this.configTracked = config;
```

Comment:
You may want to consider a null guard clause here so that the contract for non-null values is explicit.